lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -800,3 +1095,27 @@ def get_snapshot_status(self, response: Dict[str, Any] | None) -> BackupServiceS
         if "FAILED" in r_str:
             return BackupServiceState.SNAPSHOT_FAILED_UNKNOWN
         return BackupServiceState.SUCCESS
+
+
+class OpenSearchBackupFactory:
+    """Implements the logic that returns the correct class according to the cluster type.
+
+    This class is solely responsible for the creation of the correct S3 client manager.
+
+    If this cluster is an orchestrator or failover cluster, then return the OpenSearchBackup.
+    Otherwise, return the OpenSearchNonOrchestratorBackup.
+
+    There is also the condition where the deployment description does not exist yet. In this
+    case, return the base class OpenSearchBackupBase. This class solely defers all s3-related
+    events until the deployment description is available and the actual S3 object is allocated.
+    """
+
+    def __new__(cls, charm, *args, **kwargs):
```

Comment:
I would prefer that we're clear here and make it into a static method `.create()` or similar. Instantiating a factory is not very intuitive